### PR TITLE
Better integration tests discovery

### LINKED ISSUE #326: migrate.py misses migration of aliased integration tests
Some integration tests are not named after their modules, and instead have the module names listed in their `aliases` file. One such example is the `aws_lambda` integration test, which lists the `lambda` and `lambda_info` modules in its `aliases` file.

The migration script should recognize these aliases so that the tests can be correctly migrated along with the modules they are testing.

### LINKED ISSUE #327: migrate.py misses migration of plugin integration tests
Some plugins have dedicated integration tests. These should be migrated along with the plugins. The naming convention is `{plugin_type}_{plugin_name}`. Here are a few examples:

- connection_aws_ssm: lib/ansible/plugins/connection/aws_ssm.py
- lookup_hashi_vault: lib/ansible/plugins/lookup/hashi_vault.py
- filter_random_mac: lib/ansible/plugins/filter/random_mac.py

### PR DESCRIPTION
**Pull request by mkrizek**
Fixes #326
Fixes #327

### CODE REVIEW

File: migrate.py [hunk 1]
```diff
@@ -1659,21 +1659,66 @@ def integration_tests_add_to_deps(collection, dep_collection):
     integration_tests_deps.add(dep_collection)
 
 
-def poor_mans_integration_tests_discovery(checkout_dir, plugin_type, plugin_name):
-    # FIXME this might be actually enough for modules integration tests, at least for the most part
-    if plugin_type != 'modules':
-        return []
+def integration_tests_discovery(checkout_dir, plugin_type, plugin_name):
```
Comment (webknjaz) [inline migrate.py]: Let's rename this to
```suggestion
def discover_integration_tests(checkout_dir, plugin_type, plugin_name):
```

File: migrate.py [hunk 2]
```diff
@@ -1659,21 +1659,66 @@ def integration_tests_add_to_deps(collection, dep_collection):
...
+    targets_dir = os.path.join(checkout_dir, 'test/integration/targets')
+    # 'cloud/amazon/ec2_eip.py' -> 'ec2_eip'
+    plugin_name_base = os.path.basename(os.path.splitext(plugin_name)[0])
+    integration_tests_files = glob.glob(os.path.join(targets_dir, plugin_name_base))
```
Comment (webknjaz) [inline migrate.py]: do we need `recursive=True` here?

File: migrate.py [hunk 3]
```diff
@@ -1659,21 +1659,66 @@ def integration_tests_add_to_deps(collection, dep_collection):
...
+    # test/integration/targets/filter_random_mac
+    integration_tests_files.extend(glob.glob(os.path.join(targets_dir, f'{plugin_type}_{plugin_name_base}')))
```
Comment (webknjaz) [inline migrate.py]: and maybe recursive here?

File: migrate.py [hunk 4]
```diff
@@ -1659,21 +1659,66 @@ def integration_tests_add_to_deps(collection, dep_collection):
...
     deps = []
     for fname, dummy_to_remove in files:
         logger.info('Found integration tests for %s %s in %s', plugin_type, plugin_name, fname)
         deps.extend(process_needs_target(checkout_dir, fname))
 
     return files + deps
 
+processed_aliases = None
+
+def get_processed_aliases(checkout_dir):
```
Comment (webknjaz) [inline migrate.py]: No need for a global var. Just use built-in cache:


```suggestion
from functools import lru_cache

@lru_cache
def get_processed_aliases(checkout_dir):
```

File: migrate.py [hunk 5]
```diff
@@ -1659,21 +1659,66 @@ def integration_tests_add_to_deps(collection, dep_collection):
...
+    targets_dir = os.path.join(checkout_dir, 'test/integration/targets')
+    for dirpath, dirnames, filenames in os.walk(targets_dir):
+        for target in dirnames:
+            target_dir = os.path.join(targets_dir, target)
+            aliases_file = os.path.join(target_dir, 'aliases')
+            if not os.path.exists(aliases_file):
+                continue
+            for line in read_text_from_file(aliases_file).splitlines():
+                line = line.strip()
+                if not line or any(True for ignored_alias in ignored_alias_patterns if ignored_alias in line):
```
Comment (webknjaz) [inline migrate.py]: if-clause is overkill here:
```suggestion
                if not line or any(ignored_alias in line for ignored_alias in ignored_alias_patterns):
```

File: migrate.py [hunk 6]
```diff
@@ -1659,21 +1659,66 @@ def integration_tests_add_to_deps(collection, dep_collection):
...
+        'destructive',
+        'hidden',
+        'disabled',
+        'unstable',
+        'unsupported',
+    })
+    res = defaultdict(set)
+    targets_dir = os.path.join(checkout_dir, 'test/integration/targets')
+    for dirpath, dirnames, filenames in os.walk(targets_dir):
+        for target in dirnames:
```
Comment (webknjaz) [inline migrate.py]:
```suggestion
        if 'aliases' not in filenames:
            continue
        for target in dirnames:
```

File: migrate.py [hunk 7]
```diff
@@ -1659,21 +1659,66 @@ def integration_tests_add_to_deps(collection, dep_collection):
...
+            target_dir = os.path.join(targets_dir, target)
```
Comment (webknjaz) [inline migrate.py]:
```suggestion
            target_dir = dirpath
```

File: migrate.py [hunk 8]
```diff
@@ -1659,21 +1659,66 @@ def integration_tests_add_to_deps(collection, dep_collection):
...
+        '/',
+        'destructive',
+        'hidden',
+        'disabled',
+        'unstable',
+        'unsupported',
+    })
+    res = defaultdict(set)
+    targets_dir = os.path.join(checkout_dir, 'test/integration/targets')
+    for dirpath, dirnames, filenames in os.walk(targets_dir):
```
Comment (webknjaz) [inline migrate.py]: alternatively, could use something like
```suggestion
    for alias_file_name in glob.glob(os.path.join(targets_dir, '*', 'aliases')):
```

File: migrate.py (hunk 3 above)
Comment (mkrizek) [reply migrate.py]: We're looking at target names.

**Pull request merged**